File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/teleport/TeleportResult.java
package top.likoslupus.cellulosesz.api.teleport;

import top.likoslupus.cellulosesz.api.text.LocalizedMessage;
import top.likoslupus.cellulosesz.api.text.MessageArguments;

import java.util.Optional;

import static java.util.Objects.requireNonNull;

public record TeleportResult(
        TeleportStatus status,
        Optional<CellLocation> destination,
        LocalizedMessage message
) {

    public TeleportResult {
        requireNonNull(status, "status");
        requireNonNull(destination, "destination");
        requireNonNull(message, "message");
        if (status == TeleportStatus.SUCCESS && destination.isEmpty()) {
            throw new IllegalArgumentException("Successful teleport must expose its destination");
        }
    }

    public static TeleportResult success(CellLocation location) {
        return new TeleportResult(
                TeleportStatus.SUCCESS,
                Optional.of(requireNonNull(location, "location")),
                LocalizedMessage.of("service.teleport.success")
        );
    }

    public static TeleportResult failed(TeleportStatus status, String key) {
        if (status == TeleportStatus.SUCCESS) {
            throw new IllegalArgumentException("failure status required");
        }
        return new TeleportResult(
                status,
                Optional.empty(),
                LocalizedMessage.of(key)
        );
    }

    public static TeleportResult failed(
            TeleportStatus status,
            String key,
            MessageArguments arguments
    ) {
        if (status == TeleportStatus.SUCCESS) {
            throw new IllegalArgumentException("failure status required");
        }
        return new TeleportResult(
                status,
                Optional.empty(),
                LocalizedMessage.of(key, arguments)
        );
    }

    public boolean success() {
        return status == TeleportStatus.SUCCESS;
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/teleport/TeleportResult.kt
package top.likoslupus.cellulosesz.api.teleport

import top.likoslupus.cellulosesz.api.text.LocalizedMessage
import top.likoslupus.cellulosesz.api.text.MessageArguments

@JvmRecord
public data class TeleportResult(
    public val status: TeleportStatus,
    public val destination: CellLocation?,
    public val message: LocalizedMessage
) {

    init {
        require(!(status == TeleportStatus.SUCCESS && destination == null)) {
            "Successful teleport must expose its destination"
        }
    }

    public fun success(): Boolean = status == TeleportStatus.SUCCESS

    public companion object {

        @JvmStatic
        public fun success(location: CellLocation): TeleportResult =
            TeleportResult(
                TeleportStatus.SUCCESS,
                location,
                LocalizedMessage.of("service.teleport.success")
            )

        @JvmStatic
        public fun failed(status: TeleportStatus, key: String): TeleportResult {
            require(status != TeleportStatus.SUCCESS) {
                "failure status required"
            }
            return TeleportResult(status, null, LocalizedMessage.of(key))
        }

        @JvmStatic
        public fun failed(
            status: TeleportStatus,
            key: String,
            arguments: MessageArguments
        ): TeleportResult {
            require(status != TeleportStatus.SUCCESS) {
                "failure status required"
            }
            return TeleportResult(status, null, LocalizedMessage.of(key, arguments))
        }

    }

}
